File: matlab/scaling_law.m
function scaling_law()
% SCALING_LAW  Where does near-field range information actually come from?
%
% Self-contained MATLAB counterpart of sim/scaling_law.py.  Run with no
% arguments; it prints the exponent table, the closed-form ratio check, and
% the crossover range, then draws two figures.
%
% Two mutually exclusive sources of range information exist in this system.
%
%   CURVATURE  - what Luo2024, Zhang2026 and arXiv:2603.16390 all use.
%     Range enters only through the quadratic Fresnel term x^2 cos^2(th)/(2r).
%     With the complex amplitude alpha unknown, the constant part of dPhi/dr is
%     absorbed by arg(alpha), leaving only the deviation of x^2 about its mean.
%     For x uniform on [-D/2, D/2], sum_n (x^2-<x^2>)^2 = N D^4/180, hence
%         J_r^curv  = 2 SNR (2 pi fc/c)^2 cos^4(th) N D^4 / (720 r^4)
%
%   DELAY  - what nobody in this lineage uses.
%     Phi(n,m) = -(2 pi f_m/c) r_n is exactly linear in m, with no near-field
%     approximation of any kind.  One shared alpha absorbs only the mean over
%     frequency, leaving the deviation f_m - fc, hence
%         J_r^delay = 2 SNR N M (2 pi/c)^2 W^2 / 12
%
% Therefore
%     RMSE_r^curv  ~ lambda r^2 / (N^2.5 d^2 cos^2 th)     -- grows as r^2
%     RMSE_r^delay ~ c / (W sqrt(N M))                     -- flat in r and th
%     ratio        = sqrt(60 M) (W/fc) r^2 / (D^2 cos^2 th)
%
% Zhang2026 fuses subcarriers incoherently (its Eq. 44), buying sqrt(M) of
% averaging but no delay term, so it lands a factor sqrt(M) below the curvature
% bound and still sqrt(60)(W/fc) r^2/(D^2 cos^2 th) above the delay bound.

clc;
cfg = config();
r0 = 30.0;  th0 = deg2rad(15.0);  snr = 0.0;

fprintf('reference: N=%d, M=%d, W=%.0f GHz, fc=%.0f GHz, D=%.4f m, SNR=%.0f dB/element\n\n', ...
    cfg.N, cfg.M, cfg.W/1e9, cfg.fc/1e9, cfg.D, snr);

%% 1. Exponents
fprintf('%s\n1. EXPONENTS  (numerically fitted vs analytically predicted)\n%s\n', ...
    repmat('=',1,78), repmat('=',1,78));

Ns = [64 128 256 512 1024];
cv = arrayfun(@(n) rmse_curv (setfield(config(),'N',n), r0, th0, snr), Ns); %#ok<*SFLD>
dl = arrayfun(@(n) rmse_delay(setfield(config(),'N',n), r0, th0, snr, 256), Ns);
rows = {'RMSE_curv  vs N', slope(Ns,cv), -2.5; 'RMSE_delay vs N', slope(Ns,dl), -0.5};

rs = [10 20 30 50 80];
rows(end+1,:) = {'RMSE_curv  vs r', slope(rs, arrayfun(@(r) rmse_curv (cfg,r,th0,snr), rs)), 2.0};
rows(end+1,:) = {'RMSE_delay vs r', slope(rs, arrayfun(@(r) rmse_delay(cfg,r,th0,snr,256), rs)), 0.0};

Ws = [0.5 1 2 3 6]*1e9;
rows(end+1,:) = {'RMSE_delay vs W', slope(Ws, arrayfun(@(w) rmse_delay(rebuild(setfield(config(),'W',w)), r0,th0,snr,256), Ws)), -1.0};

Ms = [256 512 1024 2048];
rows(end+1,:) = {'RMSE_delay vs M', slope(Ms, arrayfun(@(m) rmse_delay(rebuild(setfield(config(),'M',m)), r0,th0,snr,128), Ms)), -0.5};
rows(end+1,:) = {'RMSE_incoh vs M', slope(Ms, arrayfun(@(m) rmse_incoh(rebuild(setfield(config(),'M',m)), r0,th0,snr,128), Ms)), -0.5};

ths = deg2rad([5 15 30 45 55]);
rows(end+1,:) = {'RMSE_curv vs 1/cos^2', slope(1./cos(ths).^2, arrayfun(@(t) rmse_curv(cfg,r0,t,snr), ths)), 1.0};

fprintf('%-22s%10s%12s\n','quantity','fitted','predicted');
for i = 1:size(rows,1)
    ok = 'OK'; if abs(rows{i,2}-rows{i,3}) > 0.06, ok = 'MISMATCH'; end
    fprintf('%-22s%10.3f%12.1f   %s\n', rows{i,1}, rows{i,2}, rows{i,3}, ok);
end

%% 2. Closed-form ratio
fprintf('\n%s\n2. THE CLOSED-FORM RATIO  vs numerically evaluated bounds\n%s\n', ...
    repmat('=',1,78), repmat('=',1,78));
fprintf('%-16s%12s%12s%11s%11s%7s\n','(r, theta)','RMSE_curv','RMSE_delay','measured','predicted','err');
for r = [10 30 50 80]
    for thd = [15 45]
        th = deg2rad(thd);
        a = rmse_curv(cfg,r,th,snr);  b = rmse_delay(cfg,r,th,snr,256);
        meas = a/b;  pred = pred_ratio(cfg,r,th);
        fprintf('(%2.0f m, %2.0f deg)   %12.3e%12.3e%11.0f%11.0f%6.1f%%\n', ...
            r, thd, a, b, meas, pred, 100*abs(meas-pred)/pred);
    end
end

%% 3. Crossover
fprintf('\n%s\n3. CROSSOVER: beyond what range is curvature ranging already hopeless?\n%s\n', ...
    repmat('=',1,78), repmat('=',1,78));
for thd = [0 15 30 45 60]
    rx = cross_range(cfg, deg2rad(thd));
    fprintf('  theta = %4.0f deg  ->  curvature competitive only within r < %.1f cm\n', thd, rx*100);
end
fprintf('\n  Zhang2026 declares its sensing region to start at %.0f m, which is %.0fx\n', ...
    15, 15/cross_range(cfg,0));
fprintf('  beyond that crossover.\n');

%% Figures
rr = linspace(1, 80, 200);
figure('Position',[100 100 1100 420]);
subplot(1,2,1);
loglog(rr, arrayfun(@(r) rmse_curv (cfg,r,th0,snr), rr), 'LineWidth', 1.8); hold on;
loglog(rr, arrayfun(@(r) rmse_incoh(cfg,r,th0,snr,256), rr), '--', 'LineWidth', 1.8);
loglog(rr, arrayfun(@(r) rmse_delay(cfg,r,th0,snr,256), rr), '-.', 'LineWidth', 1.8);
grid on; xlabel('range r (m)'); ylabel('RMSE_r bound (m)');
legend({'curvature only (narrowband)','incoherent wideband (Zhang2026 class)','coherent delay (proposed)'}, ...
    'Location','northwest','FontSize',8);
title('(a) Range bound vs distance,  \theta = 15\circ');

subplot(1,2,2);
NN = round(logspace(log10(32), log10(2048), 24));
loglog(NN, arrayfun(@(n) rmse_curv (rebuild(setfield(config(),'N',n)), r0,th0,snr), NN), 'LineWidth', 1.8); hold on;
loglog(NN, arrayfun(@(n) rmse_delay(rebuild(setfield(config(),'N',n)), r0,th0,snr,256), NN), '-.', 'LineWidth', 1.8);
grid on; xlabel('number of antennas N'); ylabel('RMSE_r bound (m)');
legend({'curvature only  (slope -2.5)','coherent delay  (slope -0.5)'}, 'Location','southwest','FontSize',8);
title('(b) Range bound vs array size,  r = 30 m');
end

% ---------------------------------------------------------------- helpers

function c = config()
c.N = 256; c.fc = 60e9; c.W = 3e9; c.M = 2048;
c = rebuild(c);
end

function c = rebuild(c)
c.c0 = 299792458;
c.lambda = c.c0/c.fc;
c.d = c.lambda/2;
c.f0 = c.fc - c.W/2;
c.D = (c.N-1)*c.d;
c.x = ((0:c.N-1) - (c.N-1)/2)*c.d;
c.freqs = c.f0 + (0:c.M-1)*(c.W/c.M);
end

function d = dist_fresnel(r, th, x)
d = r - x*sin(th) + x.^2*cos(th)^2/(2*r);
end

function [bth, br] = crlb_nb(cfg, r, th, snr_db, marg)
% Narrowband bound at fc.  marg=true marginalises the unknown complex alpha,
% which is what Zhang2026 Appendix B declares but never actually does.
k = 2*pi*cfg.fc/cfg.c0;  x = cfg.x(:);
dth = k*(x*cos(th) - x.^2*sin(2*th)/(2*r));
dr  = -k*(1 - x.^2*cos(th)^2/(2*r^2));
a   = exp(-1j*2*pi*cfg.fc*dist_fresnel(r,th,x)/cfg.c0);
D   = [a.*1j.*dth, a.*1j.*dr];
if marg, D = D - a*((a'*D)/(a'*a)); end
J   = 2*10^(snr_db/10)*real(D'*D);
Ci  = inv(J);  bth = sqrt(Ci(1,1));  br = sqrt(Ci(2,2));
end

function [bth, br] = crlb_wb(cfg, r, th, snr_db, nsub, coherent)
% Wideband bound.  coherent=true  -> one alpha shared across subcarriers, so
%                                    the phase-vs-frequency slope is observable.
% coherent=false -> an independent alpha_m per subcarrier, which is the limit
%                   of per-subcarrier processing such as Zhang2026 Eq. (41)/(44).
step = max(1, floor(cfg.M/nsub));
f = cfg.freqs(1:step:end);  f = f(1:min(nsub,numel(f)));  f = f(:);
scale = cfg.M/numel(f);
x = cfg.x(:).';                                  % 1 x N
k = 2*pi*f/cfg.c0;                               % F x 1
dth = k.*(x*cos(th) - x.^2*sin(2*th)/(2*r));     % F x N
dr  = -k.*(1 - x.^2*cos(th)^2/(2*r^2));
A   = exp(-1j*2*pi*f*dist_fresnel(r,th,x)/cfg.c0);
Dth = A.*1j.*dth;  Dr = A.*1j.*dr;
snr = 10^(snr_db/10);
if coherent
    a = A(:);  D = [Dth(:), Dr(:)];
    D = D - a*((a'*D)/(a'*a));
    J = 2*snr*real(D'*D)*scale;
else
    J = zeros(2);
    for i = 1:numel(f)
        a = A(i,:).';  D = [Dth(i,:).', Dr(i,:).'];
        D = D - a*((a'*D)/(a'*a));
        J = J + 2*snr*real(D'*D);
    end
    J = J*scale;
end
Ci = inv(J);  bth = sqrt(Ci(1,1));  br = sqrt(Ci(2,2));
end

function v = rmse_curv(cfg, r, th, snr),        [~,v] = crlb_nb(cfg,r,th,snr,true);          end
function v = rmse_delay(cfg, r, th, snr, ns),   [~,v] = crlb_wb(cfg,r,th,snr,ns,true);       end
function v = rmse_incoh(cfg, r, th, snr, ns),   [~,v] = crlb_wb(cfg,r,th,snr,ns,false);      end

function s = slope(xs, ys), p = polyfit(log(xs(:)), log(ys(:)), 1); s = p(1); end

function q = pred_ratio(cfg, r, th)
q = sqrt(60*cfg.M)*(cfg.W/cfg.fc)*r^2/(cfg.D^2*cos(th)^2);
end

function rx = cross_range(cfg, th)
rx = cfg.D*cos(th)*sqrt(cfg.fc/cfg.W/sqrt(60*cfg.M));
end

File: sim/scaling_law.py
"""
The scaling laws behind the attack: where does range information come from?

Two mutually exclusive sources of range information exist in this system.

  CURVATURE  (what Luo2024, Zhang2026 and arXiv:2603.16390 all use).
    Range enters only through the quadratic Fresnel term x^2 cos^2(theta)/(2r).
    With the complex amplitude alpha unknown, the constant part of dPhi/dr is
    absorbed by arg(alpha), leaving only the deviation of x^2 about its mean:

        dPhi_n/dr - mean  =  (2 pi f_c / c) (cos^2 th / 2r^2) (x_n^2 - <x^2>)

    For x uniform on [-D/2, D/2],  sum_n (x^2 - <x^2>)^2 = N D^4 / 180, so

        J_r^curv  =  2 SNR (2 pi f_c/c)^2 cos^4(th) N D^4 / (720 r^4)

  DELAY  (what nobody in this lineage uses).
    Phi(n,m) = -(2 pi f_m/c) r_n is exactly linear in m.  A single shared alpha
    absorbs only the mean over f, leaving the deviation f_m - f_c:

        J_r^delay =  2 SNR N M (2 pi/c)^2 W^2 / 12

Consequences, and the reason the attack is structural rather than incidental:

    RMSE_r^curv   ~  lambda r^2 / (N^2.5 d^2 cos^2 th)      grows as r^2
    RMSE_r^delay  ~  c / (W sqrt(N M))                      independent of r, th

    RMSE_curv / RMSE_delay  =  sqrt(60 M) (W/f_c) r^2 / (D^2 cos^2 th)

Zhang2026 fuses M subcarriers incoherently (Eq. 44), which buys sqrt(M) of
averaging but no delay term, so it sits a factor sqrt(M) below the curvature
bound and still sqrt(60)(W/f_c) r^2/(D^2 cos^2 th) above the delay bound.

This script checks every exponent above against the numerically evaluated FIM.
"""
import sys, os
sys.path.insert(0, os.path.dirname(__file__))
import numpy as np
from nf_model import Config, C
from estimators import crlb, crlb_wideband

SNR = 0.0


def rmse_curv(cfg, r, th):
    """Narrowband, amplitude marginalised: range from wavefront curvature only."""
    return crlb(cfg, r, th, SNR, cfg.fc, marginalize=True)[1]


def rmse_delay(cfg, r, th, n_sub=256):
    """Coherent wideband: the delay term dominates."""
    return crlb_wideband(cfg, r, th, SNR, n_sub=n_sub, coherent=True)[1]


def rmse_incoh(cfg, r, th, n_sub=256):
    """Per-subcarrier amplitudes: the class Zhang2026's Eq. (44) belongs to."""
    return crlb_wideband(cfg, r, th, SNR, n_sub=n_sub, coherent=False)[1]


def slope(xs, ys):
    return np.polyfit(np.log(xs), np.log(ys), 1)[0]


def predicted_ratio(cfg, r, th):
    D = cfg.aperture
    return np.sqrt(60 * cfg.M) * (cfg.W / cfg.fc) * r ** 2 / (D ** 2 * np.cos(th) ** 2)


def main():
    base = Config()
    r0, th0 = 30.0, np.deg2rad(15.0)
    print(f"reference: N={base.N}, M={base.M}, W={base.W/1e9:.0f} GHz, "
          f"fc={base.fc/1e9:.0f} GHz, D={base.aperture:.4f} m, SNR={SNR:.0f} dB/element\n")

    print("=" * 78)
    print("1. EXPONENTS  (numerically fitted vs analytically predicted)")
    print("=" * 78)
    rows = []

    Ns = np.array([64, 128, 256, 512, 1024])
    rows.append(("RMSE_curv  vs N", slope(Ns, [rmse_curv(Config(N=n), r0, th0) for n in Ns]), -2.5))
    rows.append(("RMSE_delay vs N", slope(Ns, [rmse_delay(Config(N=n), r0, th0) for n in Ns]), -0.5))

    rs = np.array([10., 20., 30., 50., 80.])
    rows.append(("RMSE_curv  vs r", slope(rs, [rmse_curv(base, r, th0) for r in rs]), +2.0))
    rows.append(("RMSE_delay vs r", slope(rs, [rmse_delay(base, r, th0) for r in rs]), 0.0))

    Ws = np.array([0.5e9, 1e9, 2e9, 3e9, 6e9])
    rows.append(("RMSE_delay vs W", slope(Ws, [rmse_delay(Config(W=w), r0, th0) for w in Ws]), -1.0))

    Ms = np.array([256, 512, 1024, 2048])
    rows.append(("RMSE_delay vs M", slope(Ms, [rmse_delay(Config(M=m), r0, th0, n_sub=128) for m in Ms]), -0.5))
    rows.append(("RMSE_incoh vs M", slope(Ms, [rmse_incoh(Config(M=m), r0, th0, n_sub=128) for m in Ms]), -0.5))

    print(f"{'quantity':<20}{'fitted':>10}{'predicted':>12}   {'':<6}")
    for name, fit, pred in rows:
        ok = "OK" if abs(fit - pred) < 0.06 else "MISMATCH"
        print(f"{name:<20}{fit:>10.3f}{pred:>12.1f}   {ok}")

    ths = np.deg2rad(np.array([5., 15., 30., 45., 55.]))
    fit = slope(1 / np.cos(ths) ** 2, [rmse_curv(base, r0, t) for t in ths])
    print(f"{'RMSE_curv vs 1/cos^2':<20}{fit:>10.3f}{1.0:>12.1f}   "
          f"{'OK' if abs(fit - 1) < 0.06 else 'MISMATCH'}")

    print("\n" + "=" * 78)
    print("2. THE CLOSED-FORM RATIO  vs numerically evaluated bounds")
    print("=" * 78)
    print(f"{'(r, theta)':<16}{'RMSE_curv':>12}{'RMSE_delay':>12}{'measured':>11}"
          f"{'predicted':>11}{'err':>7}")
    for r in (10., 30., 50., 80.):
        for th_deg in (15., 45.):
            th = np.deg2rad(th_deg)
            a, b = rmse_curv(base, r, th), rmse_delay(base, r, th)
            meas, pred = a / b, predicted_ratio(base, r, th)
            print(f"({r:.0f} m, {th_deg:.0f} deg){'':<3}{a:>12.3e}{b:>12.3e}"
                  f"{meas:>11.0f}{pred:>11.0f}{100*abs(meas-pred)/pred:>6.1f}%")

    print("\n" + "=" * 78)
    print("3. CROSSOVER: beyond what range is curvature ranging already hopeless?")
    print("=" * 78)
    for th_deg in (0., 15., 30., 45., 60.):
        th = np.deg2rad(th_deg)
        r_x = base.aperture * np.cos(th) * (base.fc / base.W / np.sqrt(60 * base.M)) ** 0.5
        print(f"  theta = {th_deg:4.0f} deg  ->  curvature competitive only within "
              f"r < {r_x*100:.1f} cm")
    print(f"\n  The declared sensing region of Zhang2026 starts at {base.r_min} m, "
          f"which is\n  {base.r_min/ (base.aperture*(base.fc/base.W/np.sqrt(60*base.M))**0.5):.0f}x "
          f"beyond that crossover.")


if __name__ == "__main__":
    main()
